File: server/src/algo2_py/container.py
from box import set_position, unset_position, get_size
from validations import validate_boxes_location

def get_score(box, p, boxes, container):
    set_position(box, p)
    isValid = validate_boxes_location(box, boxes, container)
    unset_position(box)

    if not isValid:
        return [0, 0]

    # Score calculation based on distance between FLB and container.
    scoreZ = container['length'] - p[2]
    scoreY = container['height'] - p[1]

    return [scoreZ, scoreY]

def update_pps(box, p, pp, container):
    pp.remove(p)

    boxSize = get_size(box)

    if boxSize['height'] + p[1] < container['height']:
        pp.add((p[0], p[1] + boxSize['height'], p[2], p[3]))

    if p[3] == 1:
        if p[0] + boxSize["width"] < container["width"]:
            pp.add((p[0] + boxSize["width"], p[1], p[2], p[3])) # b_FRB

        if p[2] + boxSize["length"] < container["length"]:
            pp.add((p[0], p[1], p[2] + boxSize["length"], p[3])) # b_RLB
    else:
        if p[0] - boxSize["width"] > 0:
            pp.add((p[0] - boxSize["width"], p[1], p[2], p[3])) # b_FLB

        if p[2] + boxSize["length"] < container["length"]:
            pp.add((p[0], p[1], p[2] + boxSize["length"], p[3])) # b_RRB

File: server/src/algorithm/algo_js/container.js
import { setPosition, unsetPosition, getBoxRotatedSize } from "./box.js";
import { validateBoxesLocation } from "./validations.js";

export const getScore = (box, p, boxes, container) => {
	setPosition(box, p);

	const isValid = validateBoxesLocation(box, boxes, container);

	unsetPosition(box, p);

	if (!isValid) {
		return [0, 0];
	}

	// Score calculation based on distance between FLB and container.
	const scoreZ = container.length - p.z;
	const scoreY = container.height - p.y;

	return [scoreZ, scoreY];
};

export const updatePps = (box, p, pp, container) => {
	pp.delete(p);

	const boxSize = getBoxRotatedSize(box);

	if (boxSize.height + p.y < container.height) {
		pp.add({
			...p,
			y: p.y + boxSize.height,
		});
	}

	if (p.dir == 1) {
		if (p.x + boxSize.width < container.width) {
			pp.add({
				...p,
				x: p.x + boxSize.width,
			});
		}
		if (p.z + boxSize.length < container.length) {
			pp.add({
				...p,
				z: p.z + boxSize.length,
			});
		}
	} else {
		if (p.x - boxSize.width > 0) {
			pp.add({
				...p,
				x: p.x - boxSize.width,
			});
		}
		if (p.z + boxSize.length < container.length) {
			pp.add({
				...p,
				z: p.z + boxSize.length,
			});
		}
	}
};
